src/VisualStudio/Core/Def/SampleToolWindow/SampleToolboxUserControl.xaml.cs

```diff
@@ -0,0 +1,425 @@
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.LanguageServer;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.ComponentModelHost;
+using Microsoft.VisualStudio.Editor;
+using Microsoft.VisualStudio.LanguageServer.Client;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.OLE.Interop;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.TextManager.Interop;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.VisualStudio.LanguageServices
+{
+    using Workspace = Microsoft.CodeAnalysis.Workspace;
+
+    /// <summary>
+    /// Interaction logic for SampleToolboxUserControl.xaml
+    /// </summary>
+    internal partial class SampleToolboxUserControl : UserControl, IOleCommandTarget
+    {
+        public SampleToolboxUserControl(Workspace workspace, IDocumentTrackingService documentTrackingService, ILanguageServiceBroker2 languageServiceBroker, IThreadingContext threadingContext)
+        {
+            InitializeComponent();
+            InitializeIfNeeded(workspace, documentTrackingService, languageServiceBroker, threadingContext);
+        }
+
+        private Workspace? workspace { get; set; }
+        private DocumentId? lastDocumentId { get; set; }
+
+        [MemberNotNullWhen(true, nameof(symbolsTreeItemsSource))]
+        private bool symbolTreeInitialized { get; set; }
+        private ObservableCollection<DocumentSymbolViewModel>? symbolsTreeItemsSource { get; set; }
+
+        private ITextSnapshot? snapshot { get; set; }
+        private IWpfTextView? textView { get; set; }
+
+        private void InitializeIfNeeded(Workspace workspace, IDocumentTrackingService documentTrackingService, ILanguageServiceBroker2 languageServiceBroker, IThreadingContext threadingContext)
+        {
+            this.workspace = workspace;
+            documentTrackingService.ActiveDocumentChanged += DocumentTrackingService_ActiveDocumentChanged;
+
+            this.textView = GetActiveTextView();
+            if (this.textView is not null)
+            {
+                this.textView.Caret.PositionChanged += FollowCursor;
+            }
+
+            searchBox.GotFocus += RemoveText;
+            searchBox.LostFocus += AddText;
+
+            void DocumentTrackingService_ActiveDocumentChanged(object sender, DocumentId? documentId)
+            {
+                threadingContext.JoinableTaskFactory.RunAsync(async () =>
+                {
+                    this.symbolTreeInitialized = false;
+                    if (documentId == this.lastDocumentId)
+                    {
+                        return;
+                    }
+
+                    this.lastDocumentId = documentId;
+                    var document = workspace.CurrentSolution.GetDocument(documentId);
```

Comment:
instead of doing all this work with the Workspace you can just get the file path directly from the IWpfTextView like this

```csharp
static string? GetFilePath(IWpfTextView textView)
{
    ThreadHelper.ThrowIfNotOnUIThread();
    if (textView.TextBuffer.Properties.TryGetProperty(typeof(IVsTextBuffer), out IVsTextBuffer bufferAdapter) &&
        bufferAdapter is IPersistFileFormat persistFileFormat &&
        ErrorHandler.Succeeded(persistFileFormat.GetCurFile(out var filePath, out _)))
    {
        return filePath;
    }

    return null;
}
```

If do that then you should be able to remove the need for `Document` or `Workspace` apis.

Reply:
The `DocumentSymbolsRequestAsync` could then just take a string (the file path) instead of a document.

LSP call would look something like this.
```csharp
var parameterFactory = new DocumentSymbolParams()
{
    TextDocument = new TextDocumentIdentifier()
    {
        Uri = new Uri(filePath)
    }
};
```